### ISSUE TITLE
Generic listify function

### ISSUE THREAD

**Issue by gabriele-marino** [microsoft/pyright#10673]
I am using pyright in strict mode and I would like to implement a "listify" function that works like the following one. I tried many different things but I keep getting errors like the one shown here. It looks like pyright isn't able to narrow the type of the input in a smart way. Is this expected? How can I solve this issue, and write a pyright-approved listify function?

```python
from collections.abc import Sequence
from typing import TypeVar

T = TypeVar("T")


def listify(x: T | Sequence[T] | None) -> list[T]:
    if x is None:
        return []
    if isinstance(x, Sequence):
        return list(x) # Argument type is "Sequence[Unknown]* | Sequence[T@listify]"
    return [x]
```

**Comment by erictraut** [comment]
Pyright's narrowing logic here aligns with type theory, so this isn't really a question of "making the narrowing logic smarter". Type theory tells us that an `isinstance` check narrows the type of the subject expression (the first argument passed to `isinstance`) to a type by intersecting it with the type specified as the second argument to `isinstance`. 

In your example, the type of the subject expression (`x`) prior to the `isinstance` check is `T | Sequence[T]`. The narrowed type becomes `(T | Sequence[T]) & Sequence[Unknown]`. Here I'm using `&` to represent an intersection operator. The concept of an intersection may be new to you because the Python type system doesn't yet expose this concept. (There's ongoing work to add this.) We can further simplify this type by distributing the intersection: `(T & Sequence[Unknown]) | (Sequence[T] & Sequence[Unknown])`, and this further simplifies to `(T & Sequence[Unknown]) | Sequence[T]`.

Internally, pyright does not fully implement support for intersections, but it does model intersections between concrete types and type variables using a feature called [conditional types](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=conditional-types-and-type-variables). The type `T & Sequence[Unknown]` is displayed as `Sequence[Unknown]*`. If you look at the revealed type of `x` after the `isinstance` conditional, pyright reports its narrowed type as `Sequence[Unknown]* | Sequence[T]`, which is consistent with type theory.

You might wonder why this theoretical behavior is necessary for correctness when a narrowed type of `Sequence[T]` is what you might expect. The problem is that this would allow unsound code to type check without any error, even in strict mode.

Here's an example.

```python
def func(x: Sequence[str] | Sequence[int]):
    y = listify(x)
    reveal_type(y) # list[str | Sequence[int]]
    print(y)

func([1]) # prints "[1]" which is not a valid "list[str | Sequence[int]]"
```

I realize that this answer may not be very satisfying because it doesn't really resolve your issue. 

One option is to suppress the strict-mode diagnostic using a `# type: ignore` or `# pyright: ignore` comment. This effectively tells pyright that you are OK with the (admittedly small) soundness hole.

Perhaps other pyright users can offer other suggestions.
